README.md
# UserStyles.org Archive

![UserStyles.org Auto Update](https://github.com/33kk/uso-archive/workflows/UserStyles.org%20Auto%20Update/badge.svg)


changelog.txt
commit e326ba9ba0ffb79b891d9b46435fcffeea057b68
Author: 33kk <goncarmarko@gmail.com>
Date:   Wed Sep 30 21:36:40 2020 +0300

    data: formatting

commit 7b80f20cac8e94a2ac095b5d8682914e25014d22
Author: 33kk <goncarmarko@gmail.com>
Date:   Wed Sep 30 20:30:52 2020 +0300

    scraper: fix commit messages saying update when style is new

commit 0581d9f81cba3e60c95ec2f263c5b18d6a662658
Author: 33kk <goncarmarko@gmail.com>
Date:   Wed Sep 30 20:29:15 2020 +0300

    scraper: store files formatted

commit 3b4a2f3992f6a07ecdc4745be55ec0e516545da0
Author: 33kk <goncarmarko@gmail.com>
Date:   Wed Sep 30 20:27:46 2020 +0300

    ci: add turnstyle action, set update interval to every 6 hours

commit d3cf27c5af06394cad9cf61ad3801f38367d6526
Author: 33kk <goncarmarko@gmail.com>
Date:   Wed Sep 30 20:27:06 2020 +0300

    converters: fix unknown id in markdown

commit f24315c96046195f94a8340b398cb9870fa76438
Author: 33kk <goncarmarko@gmail.com>
Date:   Wed Sep 30 15:36:31 2020 +0300

    web: fix screenshot margin and move author and license

commit 7289d861ecc5a3b59b4f8ef90c83f28fb52ef3a8
Author: 33kk <goncarmarko@gmail.com>
Date:   Wed Sep 30 15:35:58 2020 +0300

    web: show total installs

commit 1cf264db2e239c1a4adb48a3764e63b3f63cfcce
Author: GitHub Actions <actions@github.com>
Date:   Wed Sep 30 12:19:41 2020 +0000

    data: auto update
    [Update, 2020-09-30T08:15:27.000Z] Fiery Red theme for SkyCrypt (190135)
    [Update, 2020-09-30T11:15:29.000Z] no baners.youtube (172587)
    [Update, 2020-09-30T10:09:25.000Z] Compact Tube (White) v1.0.8 (β) (186936)
    [Update, 2020-09-30T09:43:49.000Z] ProtonMail Midnight Blue <beta> (160537)
    [Update, 2020-09-30T11:47:41.000Z] 3ddd.ru - Dark Theme (163681)
    [Update, 2020-09-30T10:17:31.000Z] Dokomi Expo (190134)
    [Update, 2020-09-30T08:27:56.000Z] NicoVideo Dark Theme Custom (187134)
    [Update, 2020-09-30T09:19:27.000Z] EZGIF Dark (190136)

commit 3ad7ef08778d10c229445b66b197acb4a8b78309
Author: 33kk <goncarmarko@gmail.com>
Date:   Wed Sep 30 15:08:28 2020 +0300

    scraper: move converters to separate file

commit f3e9af48dfabfcd538df226664b04f708c936fbb
Author: 33kk <goncarmarko@gmail.com>
Date:   Wed Sep 30 14:58:08 2020 +0300

    scraper: hide android styles from index

commit 636c0d1987ee2202ff9aa69ee86db767390f7c66
Author: 33kk <goncarmarko@gmail.com>
Date:   Wed Sep 30 14:57:23 2020 +0300

    data: hide android styles from index

commit 22366f2ac0d61bcfab96433839e84d2f05c91ee6
Author: GitHub Actions <actions@github.com>
Date:   Wed Sep 30 08:16:35 2020 +0000

    data: auto update
    [Update, 2020-09-30T05:05:13.000Z] Canta darker for Mastodon (180297)
    [Update, 2020-09-30T07:19:40.000Z] 夜间斗鱼 NightMode For Douyu.com (158117)
    [Update, 2020-09-30T07:23:44.000Z] stylish NightMode (190132)
    [Update, 2020-09-30T06:19:20.000Z] Dark theme for Google Translate (190133)
    [Update, 2020-09-30T07:59:03.000Z] NicoVideo Dark Theme Custom (187134)
    [Update, 2020-09-30T08:11:56.000Z] Dokomi Expo (190134)
    [Update, 2020-09-30T04:36:21.000Z] 百度系列改造 (185937)

commit 9bd7f9abcc898305b0e36445d3d89987f1af2068
Author: GitHub Actions <actions@github.com>
Date:   Wed Sep 30 04:16:30 2020 +0000

    data: auto update
    [Update, 2020-09-30T00:37:04.000Z] Ameblo Management (158183)
    [Update, 2020-09-30T01:42:15.000Z] 作業実施報告書 | Workship (190126)
    [Update, 2020-09-30T00:44:40.000Z] F-List WX Blue (New Site Desktop Width) (188323)
    [Update, 2020-09-30T00:45:12.000Z] F-List WX Dark (New Site Desktop Width) (188057)

commit 759b6e406de2842a3a1ef896e30c6161d4ac7eb6
Author: GitHub Actions <actions@github.com>
Date:   Wed Sep 30 00:38:47 2020 +0000

    data: auto update
    [Update, 2020-09-28T21:08:52.000Z] Compact Tube (White) v1.0.7 (β) (186936)
    [Update, 2020-09-28T22:25:16.000Z] YouTube Black - KMV (174767)
    [Update, 2020-09-29T23:47:38.000Z] Amazon - Minimal Dark (166875)
    [Update, 2020-09-29T20:20:42.000Z] 3ddd.ru - Dark Theme (163681)
    [Update, 2020-09-29T03:57:25.000Z] Bob Esponja Style for Google.com.br (190065)
    [Update, 2020-09-29T00:21:01.000Z] Wolvden Dark Blues (190052)
    [Update, 2020-09-29T04:06:49.000Z] UnderTale Chara (190072)
    [Update, 2020-09-29T04:00:03.000Z] TBBT Style for Google.com.br (190066)
    [Update, 2020-09-29T21:01:50.000Z] Dededianas (190115)
    [Update, 2020-09-28T19:59:20.000Z] Amazon - Dark slate (133725)
    [Update, 2020-09-28T13:50:29.000Z] Reactor - Anime Style (with Dark theme) (138637)
    [Update, 2020-09-29T00:43:07.000Z] Readable Twitch (190029)
    [Update, 2020-09-28T22:13:19.000Z] VK Plus (107316)
    [Update, 2020-09-29T05:54:51.000Z] BaiduXs 20年09月百度多栏 百度Xs-简洁 多栏 卡片 (189519)
    [Update, 2020-09-29T13:58:07.000Z] weibo-dark (190089)
    [Update, 2020-09-29T11:46:09.000Z] Planeta (190084)
    [Update, 2020-09-29T14:17:34.000Z] Dark & section coms pour gogos pas drôles disabled (180969)
    [Update, 2020-09-29T03:13:00.000Z] algo........ (190070)
    [Update, 2020-09-28T16:15:59.000Z] Scrollbar Custom - KMV (182441)
    [Update, 2020-09-29T23:33:53.000Z] TorrentFreak - Minimal Dark (156697)
    [Update, 2020-09-28T19:25:30.000Z] The Bell Tree - 2020 Fair Theme (189774)
    [Update, 2020-09-28T18:16:59.000Z] μTorrent Web Clear UI (190027)
    [Update, 2020-09-29T16:50:10.000Z] Yahoo Fantasy Dark Night Theme (152103)
    [Update, 2020-09-29T03:58:48.000Z] Flamengo Style for Google.com.br (190064)
    [Update, 2020-09-29T11:00:00.000Z] Clean Human Benchmark (190080)
    [Update, 2020-09-28T15:47:09.000Z] Custom KickAssAnime (173117)
    [Update, 2020-09-28T20:57:07.000Z] bubbleteabrin (190032)
    [Update, 2020-09-29T06:32:01.000Z] Google dark search (188928)
    [Update, 2020-09-29T11:13:42.000Z] Tunisianet Dark (190081)
    [Update, 2020-09-28T20:21:35.000Z] 2ch Reborn (165011)
    [Update, 2020-09-29T16:24:25.000Z] Dark Google Collection (174098)
    [Update, 2020-09-29T09:38:26.000Z] Roblox aqua (DARK MODE) (184772)

pepega.js
const fs = require("fs").promises;
const path = require("path");
const utils = require("./lib/utils");
const uso = require("./lib/uso");
const logger = require("./lib/logger");

async function main() {
	const index = {};
	const files = await fs.readdir(path.resolve(__dirname, "data", "uso-styles"));
	const downloadedScreenshots = await fs.readdir(path.resolve(__dirname, "data", "screenshots"));
	for (const stylePath of files/*["131452.json"]*/) {
		const json = JSON.parse(await fs.readFile(path.resolve(__dirname, "data", "uso-styles", stylePath)));

		const h = uso.convertFromUso(json);
		index[json.id] = h.indexStyle;

		if (h.style.screenshots) {
			if (h.style.screenshots.main) {
				h.style.screenshots.main.archived = downloadedScreenshots.includes(h.style.screenshots.main.name);
			}
			if (h.style.screenshots.additional) {
				for (const screenshot of h.style.screenshots.additional) {
					screenshot.archived = downloadedScreenshots.includes(screenshot.name);
				}
			}
		}

		uso.saveStyle({ id: h.id, style: h.style });
		
		logger.info(["Regen"], `Regenerated ${json.id}`);
	}
	
	await utils.saveIndexes(index);
}

main().catch(e => logger.error(["Regen"], e));



lib/utils.js
const fetch = require("node-fetch");
const fsSync = require("fs");
const fs = require("fs").promises;
const path = require("path");
const logger = require("./logger");

async function sleep(time) {
	return new Promise((resolve) => {
		setTimeout(resolve, time);
	});
}

async function downloadFile(url, path) {
	const res = await fetch(url);
	const fileStream = fsSync.createWriteStream(path);
	await new Promise((resolve, reject) => {
		res.body.pipe(fileStream);
		res.body.on("error", reject);
		fileStream.on("finish", resolve);
	});
}

async function retry(func, maxErrors, dontThrow, delay) {
	let errors = 0;
	for(;;) {
		try {
			return await func();
		}
		catch (e) {
			if (errors < maxErrors) {
				logger.error(["Utils"], e);
				errors++;
				if (delay) {
					await sleep(delay);
				}
			}
			else {
				if (!dontThrow)
					throw e;
				else return;
			}
		}
	}
}

async function waitUntilTrue(func, checkInterval) {
	while (!func()) {
		await sleep(checkInterval || 50);
	}
}

async function saveIndexes(index) {
	let arrayIndex = [];
	for (const i in index) {
		arrayIndex.push(index[i]);
	}
	arrayIndex.sort((a, b) => b.stats.installs.weekly - a.stats.installs.weekly);

	let searchIndex = [];
	for (const el of arrayIndex) {
		let data = {f: el.info.format, i: el.id, n: el.info.name, /*d: el.info.description, */c: el.info.category, u: Math.floor(+new Date(el.info.updatedAt) / 1000), t: el.stats.installs.total, w: el.stats.installs.weekly, r: el.stats.rating};
		if (el.info.author) {
			data.ai = el.info.author.id;
			data.an = el.info.author.name;
		}
		if (el.screenshots && el.screenshots.main) {
			data.sn = el.screenshots.main.name;
			data.sa = el.screenshots.main.archived;
		}
		searchIndex.push(data);
	}

	await fs.writeFile(path.resolve(__dirname, "..", "data", "index.json"), JSON.stringify(index), { encoding: "utf8" });
	await fs.writeFile(path.resolve(__dirname, "..", "data", "search-index.json"), JSON.stringify(searchIndex), { encoding: "utf8" });
}

module.exports = {
	sleep,
	retry,
	waitUntilTrue,
	downloadFile,
	saveIndexes
};


handle-issue.js
const uso = require("./lib/uso");
const utils = require("./lib/utils");
const logger = require("./lib/logger");
const fs = require("fs").promises;
const path = require("path");
const performance = require("perf_hooks").performance;

async function main() {
	const issueData = JSON.parse(process.env.ISSUE_DATA);
	switch (process.argv[2]) {
	case "uso":
		updateUso(issueData);
		break;
	case "usercss":
		updateUsercss(issueData);
		break;
	}
}

async function updateUso(issueData) {
	let index = JSON.parse(await fs.readFile(path.resolve(__dirname, "data", "index.json"), { encoding: "utf8" }));
	let changes = [];
	const styleIds = issueData.body.split(/, ?/).slice(0, 50);
	
	for (const styleId of styleIds) {
		const id = parseInt(styleId);
		if (Number.isNaN(id) || !id) continue;
		logger.info(["USO Update Request"], `Style ${id} downloading...`);
		const p = performance.now();
		const data = await utils.retry(async () => await uso.saveStyle(await uso.getStyle(id, true, true)), 50);
		logger.info(["USO Update Request"], `Style ${data.style.info.name} (${id}) downloaded, took ${Math.ceil((performance.now() - p) / 1000)}s`);
		changes.push({ type: index[id] === undefined ? "New" : "Update", updatedAt: data.style.info.updatedAt, name: data.style.info.name, id });
		index[id] = data.indexStyle;
	}

	logger.info(["USO Update Request"], "Saving indexes...");
	await utils.saveIndexes(index);
	console.log("::set-output name=changes::" + ["data: uso update request", ...changes.map(e => `[${e.type === "new" ? "New" : "Update"}, ${e.updatedAt}] ${e.name} (${e.id})`)].join("%0A").replace(/"/g, "\\\""));
}

async function updateUsercss(issueData) {
	console.log(issueData);
}

main();


get-style-info.js
const fs = require("fs").promises;
const path = require("path");
const utils = require("./lib/utils");
const uso = require("./lib/uso");
const logger = require("./lib/logger");

async function main() {
	let started = 0;
	const maxInParralel = 100;

	const skipTo = parseInt(process.argv[2]);
	const files = (await fs.readdir(path.resolve(__dirname, "data", "styles"))).sort().slice(skipTo);

	let regular = skipTo || 0;
	let full = 0;
	const total = (skipTo || 0) + files.length;

	for (const stylePath of files) {
		await utils.waitUntilTrue(() => started < maxInParralel);
		
		(async () => {
			started++;
			const json = JSON.parse(await fs.readFile(path.resolve(__dirname, "data", "styles", stylePath)));
		
			if (uso.shouldBeFullyArchived(json) && json.screenshots && json.screenshots.main && json.screenshots.main.archived === false && json.screenshots.main.name.includes("-")) {
				try {
					await uso.saveStyle(await uso.getStyle(json.id, true, true));
				}
				catch (e) {
					logger.error(["Get Style Info"], e);
				}
				full++;
				logger.info(["Get Style Info"], `Style ${json.id} is fully saved. Progress: ${regular+full}/${total} (full: ${full}, regular: ${regular})`);
			} else {
				// const styleData = uso.convertStyle(json);
				// await uso.saveStyle({ usercss: styleData.usercss, markdown: styleData.markdown });
				regular++;
				// logger.info(["Get Style Info"], `Style ${json.id} is saved. Progress: ${regular+full}/${total} (full: ${full}, regular: ${regular})`);
			}
			started--;
		})().catch(e => logger.error(["Get Style Info"], e));
	}

	await utils.waitUntilTrue(() => started === 0, 5000);
}

main().catch(e => logger.error(["Get Style Info"], e));



lib/converters.js
/* eslint-disable no-case-declarations */
const utils = require("./utils");

function toMarkdown(style) {
	return `# ${style.info.name} (${style.id})${style.obsolete ? `
### Obsolete
- Obsoleting style id: ${style.obsolete.obsoletedBy.id}
- Obsoleting style name: ${style.obsolete.obsoletedBy.name}
- Obsoletion message: ${style.obsolete.message}
` : ""}${style.deleteReasonId? `### Deleted
- Delete reason id: ${style.deleteReasonId}
` : ""}

### Information
- Rating: ${style.stats.rating || "unknown"}
- Applies to: ${style.info.category}
- Updated at: ${style.info.updatedAt ? style.info.updatedAt : "unknown"}
- Created at: ${style.info.createdAt ? style.info.createdAt : "unknown"}
- Weekly installs: ${style.stats.installs.weekly}
- Total installs: ${style.stats.installs.total}
- Author: ${style.info.author ? style.info.author.name || "unknown" : "unknown"} (${style.info.author ? style.info.author.id || "unknown" : "unknown"})
- License: ${style.info.license || "unknown"}

${style.info.description ? `### Description
${style.info.description}
` : ""}
${style.info.additionalInfo ? `### Update notes
${style.info.additionalInfo}
` : ""}
${style.info.additionalInfo ? `### Files
- [UserCSS](../usercss/${style.id}.user.css)
- [JSON](../styles/${style.id}.json)
` : ""}
${style.screenshots ? `### Screenshots
${style.screenshots.main ? screenshotToMarkdown(style.screenshots.main) : ""}${style.screenshots.additional && style.screenshots.additional.length > 0 ? `
${style.screenshots.additional.map(s => screenshotToMarkdown(s)).join("\n")}` : ""}` : ""}`;
}

function screenshotToMarkdown(screenshot) {
	if (screenshot.archived) {
		return markdownImg(`../screenshots/${screenshot.name}`);
	}
	else {
		return markdownImg(`https://userstyles.org/${screenshot.name.includes("-") ? "auto_" : ""}style_screenshots/${screenshot.name}`);
	}
}

function markdownImg(url) {
	return `![${url}](${url})`;
}

function toMarkdownIndexEntry(style) {
	return `# ${style.info.name} (${style.id})
- Updated at: ${style.info.updatedAt ? new Date(style.info.updatedAt).toLocaleString() : "unknown"}
- Category: ${style.info.category}
- Weekly installs: ${style.stats.installs.weekly}

${style.info.description ? `### Description
${style.info.description}
` : ""}`;
}

function toUsercss(style) {
	let replaces = [];
	const settings = toUsercssSettings(style.style.settings);

	function processKey(key) {
		if (/^[\w-_]+$/.test(key)) {
			return key;
		}
		else {
			const result = key.replace(/[^\w-_]/g, "-");
			const c = replaces.filter(v => v.r === result).length || null;
			replaces.push({ from: new RegExp(utils.escapeRegExp(`/*[[${key}]]*/`), "gi"), to: `/*[[${result}${c ? "-" + c : ""}]]*/`, r: result });
			return result;
		}
	}

	function toUsercssDropdownOptions(options) {
		let result = "";
		for (const option of options) {
			if (option.default) result = `${processKey(option.key)} "${option.label.replace(/"/g, "\\\"")}${option.default ? "*" : ""}" <<<EOT ${option.value} EOT;\r\n` + result;
			else result += `${processKey(option.key)} "${option.label.replace(/"/g, "\\\"")}${option.default ? "*" : ""}" <<<EOT ${option.value} EOT;\r\n`;
		}
		return result;
	}

	function toUsercssSetting(setting) {
		switch (setting.type) {
		case "dropdown":
			return `@advanced dropdown ${processKey(setting.key)} "${setting.label.replace(/"/g, "\\\"")}" {
	${toUsercssDropdownOptions(setting.options)}
}`;
		case "color":
			return `@advanced color ${processKey(setting.key)} "${setting.label.replace(/"/g, "\\\"")}" ${setting.options[0].value}`;
		case "image":
			const key = processKey(setting.key);
			return `@advanced dropdown ${key} "${setting.label.replace(/"/g, "\\\"")}" {
${toUsercssDropdownOptions(setting.options)}
	${processKey(key)}-custom-dropdown "Custom" <<<EOT /*[[${processKey(key)}-custom]]*/ EOT;
}
@advanced text ${processKey(key)}-custom "${setting.label.replace(/"/g, "\\\"")} (Custom)" "https://example.com/image.png"`;
		case "text":
			return `@advanced text ${processKey(setting.key)} "${setting.label.replace(/"/g, "\\\"")}" "${setting.options[0].value}"`;
		}
	}

	function toUsercssSettings(settings) {
		if (!settings) return "";
		let result = "";
		for (const setting of settings) {
			result += `${toUsercssSetting(setting).replace(/\*\//g, "*\\/")}\r\n`;
		}
		return result;
	}

	let result = `/* ==UserStyle==
@name           ${style.info.name}
@namespace      USO Archive
@author         ${style.info.author ? style.info.author.name || "unknown" : "unknown"}
@description    \`${style.info.description ? style.info.description.replace(/[\r\n]/gm, "") : "none"}\`
@version        ${versionFromTimestamp(style.info.updatedAt)}
@license        ${style.info.license}
@preprocessor   uso${settings ? "\n" + settings : ""}
==/UserStyle== */
${style.style.css}`;

	for (const replace of replaces) {
		result = result.replace(replace.from, replace.to);
	}

	return result;
}

function versionFromTimestamp(timestamp) {
	const date = new Date(timestamp);
	return `${date.getFullYear()}${date.getMonth() + 1}${date.getDate()}.${date.getHours()}.${date.getMinutes()}`;
}

function toIndexEntry(style) {
	return {
		id: style.id,
		info: {
			name: style.info.name,
			description: style.info.description,
			format: style.info.format,
			category: style.info.category,
			updatedAt: style.info.updatedAt,
			author: style.info.author
				? {
					id: style.info.author.id,
					name: style.info.author.name,
				}
				: undefined,
		},
		stats: style.stats,
		screenshots: style.screenshots ? {
			main: style.screenshots.main || undefined,
		} : undefined,
	};
}

module.exports = {
	toIndexEntry,
	toUsercss,
	toMarkdownIndexEntry,
	toMarkdown
};


lib/uso.js
const fetch = require("node-fetch");
const utils = require("./utils");
const converters = require("./converters");
const fs = require("fs").promises;
const path = require("path");

async function getPage(page, category, perPage) {
	const response = await fetch(`https://userstyles.org/api/v1/styles/${category}?per_page=${perPage}&page=${page}`, {
		headers: {
			Referer: "https://userstyles.org/",
			"User-Agent": "Mozilla/5.0 (Windows NT 10.0; rv:78.0) Gecko/20100101 Firefox/78.0",
		},
	});

	if (response.status !== 200) throw `Non 200 status code (${response.status})`;

	let json = await response.json();
	json.data.forEach((style) => {if (style.user) style.user.email = "redacted"; });
	return json;
}

async function getStyleInfo(id) {
	const response = await fetch(`https://userstyles.org/api/v1/styles/${id}`, {
		headers: {
			Referer: "https://userstyles.org/",
			"User-Agent": "Mozilla/5.0 (Windows NT 10.0; rv:78.0) Gecko/20100101 Firefox/78.0",
		},
	});
	if (response.status !== 200) throw `Non 200 status code (${response.status})`;
	let json = await response.json();
	json.user.email = "redacted";
	return json;
}

function filenameFromScreenshotUrl(url, thr) {
	if (!url) return undefined;
	const rgx = "style_screenshots\\/(?<filename>.*?)\\?";
	const match = new RegExp(rgx).exec(url);
	if (thr) {
		if (!match) throw "No matches. Input: " + url;
		if (!match.groups) throw "No groups. Input: " + url;
		if (!match.groups.filename) throw "No filename group. Input: " + url;
	} else {
		if (!match || !match.groups || !match.groups.filename) return undefined;
	}
	return match.groups.filename;
}

function licenseToSpdx(license) {
	switch (license) {
	case "publicdomain":
		return "CC0-1.0";
	case "ccby":
		return "CC-BY-4.0";
	case "ccbysa":
		return "CC-BY-SA-4.0";
	case "ccbynd":
		return "CC-BY-ND-4.0";
	case "ccbync":
		return "CC-BY-NC-4.0";
	case "ccbyncsa":
		return "CC-BY-NC-SA-4.0";
	case "ccbyncnd":
		return "CC-BY-NC-ND-4.0";
	}
	return "NO-REDISTRIBUTION";
}

function toBetterFormat(style, isScreenshotArchived, areAdditionalArchived) {
	let settings = style.style_settings
		? style.style_settings.map((s) => {
			return {
				key: s.install_key,
				label: s.label,
				type: s.setting_type,
				options: s.style_setting_options
					? s.style_setting_options.map((o) => {
						return {
							key: o.install_key,
							label: o.label,
							value: o.value,
							default: o.default,
						};
					})
					: undefined,
			};
		})
		: hasSettings(style.css)
			? null
			: undefined;
	if (Array.isArray(settings)) {
		if (settings.length === 0) {
			settings = undefined;
		}
	}

	const mainScreenshotName = filenameFromScreenshotUrl(style.after_screenshot_name || style.screenshot_url || style.screenshot || undefined);

	return {
		id: style.id,
		info: {
			name: style.name,
			description: style.description,
			additionalInfo: style.additional_info,
			format: style.category === "android" ? "uso-android" : "uso",
			category: style.category === "global" ? "global" : style.subcategory || style.raw_subcategory,
			createdAt: style.created,
			updatedAt: style.updated,
			license: licenseToSpdx(style.license),
			author: style.user
				? {
					id: style.user.id,
					name: style.user.name,
					paypalEmail: style.user.paypal_email || undefined,
					homepage: style.user.homepage || undefined,
				}
				: undefined,
		},
		stats: {
			installs: {
				total: style.total_install_count,
				weekly: style.weekly_install_count,
			},
			rating: style.rating || undefined,
		},
		screenshots: mainScreenshotName || style.screenshots ? {
			main: mainScreenshotName ? { name: mainScreenshotName, archived: !!isScreenshotArchived } : undefined,
			additional: style.screenshots ? style.screenshots.map((s) => { return { name: filenameFromScreenshotUrl(s), archived: !!areAdditionalArchived};}) : undefined,
		} : undefined,
		discussions: {
			stats: {
				discussionsCount: style.discussionsCount,
				commentsCount: style.commentsCount,
			},
			data: style.discussions
				? style.discussions.map((d) => {
					return {
						id: d.id,
						title: d.name,
						createdAt: d.created,
						author: {
							id: d.author_id,
							name: d.author_name,
						},
					};
				})
				: undefined,
		},
		obsolete: style.obsolete
			? {
				obsoletedBy: {
					id: style.obsoleting_style_id,
					name: style.obsoleting_style_name,
				},
				message: style.obsoletion_message
			}
			: undefined,
		deleteReason: style.admin_delete_reason_id ? { id: style.admin_delete_reason_id, message: "None" } : undefined,
		style: {
			css: style.css,
			settings,
		},
	};
}

async function getStyle(id, willMainBeArchived, willAdditionalBeArchived) {
	const usoStyle = await utils.retry(async () => await getStyleInfo(id), 50);
	const style = toBetterFormat(usoStyle, willMainBeArchived, willAdditionalBeArchived);
	const styleData = { ...convertStyle(style, willMainBeArchived, willAdditionalBeArchived), usoStyle };
	return styleData;
}

function convertFromUso(usoStyle, willMainBeArchived, willAdditionalBeArchived) {
	const style = toBetterFormat(usoStyle, willMainBeArchived, willAdditionalBeArchived);
	return { ...convertStyle(style), usoStyle };
}

function convertStyle(style) {
	const indexStyle = converters.toIndexEntry(style);
	const usercss = style.info.format === "uso" ? converters.toUsercss(style) : undefined;
	// const markdown = converters.toMarkdown(style);
	let screenshots = undefined;

	if (style.screenshots) {
		screenshots = [];
		if (style.screenshots.main)
			if (style.screenshots.main.name.includes("-")) {
				screenshots.push({ url: `https://userstyles.org/auto_style_screenshots/${style.screenshots.main.name}`, name: style.screenshots.main.name });
			}
			else {
				screenshots.push({ url: `https://userstyles.org/style_screenshots/${style.screenshots.main.name}`, name: style.screenshots.main.name });
			}
		if (style.screenshots.additional) {
			style.screenshots.additional.map(e => screenshots.push({ url: `https://userstyles.org/style_screenshots/${e.name}`, name: e.name }));
		}
	}
	return { id: style.id, style, indexStyle, usercss, screenshots };
}

async function saveStyle(styleData) {
	const { id, style, usoStyle, usercss, screenshots } = styleData;
	if (id) {
		if (style)
			await fs.writeFile(path.resolve(__dirname, "..", "data", "styles", `${id}.json`), JSON.stringify(style, null, 2), { encoding: "utf8" });
		if (usoStyle)
			await fs.writeFile(path.resolve(__dirname, "..", "data", "uso-styles", `${id}.json`), JSON.stringify(usoStyle, null, 2), { encoding: "utf8" });
		if (usercss)
			await fs.writeFile(path.resolve(__dirname, "..", "data", "usercss", `${id}.user.css`), usercss, { encoding: "utf8" });
		// if (markdown)
		// 	await fs.writeFile(path.resolve(__dirname, "..", "data", "markdown", `${id}.md`), markdown, { encoding: "utf8" });
	}

	if (screenshots && screenshots.length > 0) {
		for (const screenshot of screenshots) {
			await utils.retry(async () => {
				await utils.downloadFile(screenshot.url, path.resolve(__dirname, "..", "data", "screenshots", screenshot.name));
			}, 15);
		}
	}
	return styleData;
}

const placeholderRegex = /\/\*\[\[.*?\]\]\*\//gm;

function hasSettings(css) {
	return placeholderRegex.test(css);
}

function shouldBeFullyArchived(style) {
	return (style.stats.installs.total > 1000 && new Date(style.info.updatedAt) > new Date(2015, 1, 1, 0, 0, 0, 0)) || new Date(style.info.updatedAt) > new Date(2019, 1, 1, 0, 0, 0, 0) || (style.style && style.style.settings) || (style.screenshots && style.screenshots.additional && style.screenshots.additional.length > 0);
}

module.exports = {
	hasSettings,
	shouldBeFullyArchived,
	getPage,
	getStyleInfo,
	getStyle,
	convertStyle,
	convertFromUso,
	saveStyle,
	filenameFromScreenshotUrl,
	licenseToSpdx,
	toBetterFormat
};
